File: plugin-tester-scala/src/main/scala/example/myapp/helloworld/AuthenticatedGreeterClient.scala
/*
 * Copyright (C) 2018-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package example.myapp.helloworld

import akka.actor.ActorSystem
import akka.grpc.GrpcClientSettings
import example.myapp.helloworld.grpc._

import scala.concurrent.Await
import scala.concurrent.duration._

object AuthenticatedGreeterClient {
  def main(args: Array[String]): Unit = {
    // Boot akka
    implicit val sys = ActorSystem("HelloWorldClient")

    // Take details how to connect to the service from the config.
    val clientSettings = GrpcClientSettings.connectToServiceAt("127.0.0.1", 8082).withTls(false)
    // Create a client-side stub for the service
    val client: GreeterServiceClient = GreeterServiceClient(clientSettings)

    val failureWhenUnauthenticated = Await.result(client.sayHello(HelloRequest("Alice")).failed, 10.seconds)
    sys.log.warning(s"Call without authentication fails: $failureWhenUnauthenticated")

    val replyWhenAuthenticated =
      Await.result(client.sayHello().addHeader("Token", "XYZ").invoke(HelloRequest("Alice")), 10.seconds)
    sys.log.warning(s"Call with authentication succeeds: $replyWhenAuthenticated")

    val interceptedClient: GreeterServiceClient =
      client.addRequestHeader("Token", "XYZ")

    val replyWhenInterceptAuthenticated =
      Await.result(interceptedClient.sayHello(HelloRequest("Alice")), 10.seconds)
    sys.log.warning(s"Call with intercepted authentication succeeds: $replyWhenInterceptAuthenticated")

  }
}

File: plugin-tester-java/src/main/java/example/myapp/helloworld/AuthenticatedGreeterClient.java
package example.myapp.helloworld;

import akka.actor.ActorSystem;
import akka.grpc.GrpcClientSettings;
import example.myapp.helloworld.grpc.GreeterServiceClient;
import example.myapp.helloworld.grpc.HelloReply;
import example.myapp.helloworld.grpc.HelloRequest;
import io.grpc.StatusRuntimeException;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

public class AuthenticatedGreeterClient {

  public static void main(String[] args) throws Exception {

    String serverHost = "127.0.0.1";
    int serverPort = 8082;

    ActorSystem system = ActorSystem.create("HelloWorldClient");

    // Configure the client by code:
    GrpcClientSettings settings = GrpcClientSettings.connectToServiceAt(serverHost, serverPort, system).withTls(false);

    GreeterServiceClient client = null;
    try {
      client = GreeterServiceClient.create(settings, system);

      try {
        client.sayHello(HelloRequest.newBuilder().setName("Alice").build()).toCompletableFuture().get(5, TimeUnit.SECONDS);
        throw new RuntimeException("Call should have failed");
      } catch (ExecutionException ex) {
        system.log().warning(ex, "Call without authentication fails as expected");
      }

      HelloReply replyWhenAuthenticated = client.sayHello()
        .addHeader("Token", "XYZ")
        .invoke(HelloRequest.newBuilder().setName("Alice").build()).toCompletableFuture().get(5, TimeUnit.SECONDS);
      system.log().info("Call with authentication succeeds: {}", replyWhenAuthenticated);

      GreeterServiceClient clientWithMeta = client.addRequestHeader("Token", "XYZ");

      HelloReply replyWhenInterceptAuthenticated = clientWithMeta.sayHello()
        .addHeader("Token", "XYZ")
        .invoke(HelloRequest.newBuilder().setName("Alice").build()).toCompletableFuture().get(5, TimeUnit.SECONDS);
      system.log().info("Call with authentication succeeds: {}", replyWhenInterceptAuthenticated);

    } catch (StatusRuntimeException e) {
      System.out.println("Status: " + e.getStatus());
    } catch (Exception e)  {
      System.out.println(e);
    } finally {
      if (client != null) client.close();
      system.terminate();
    }
  }
}
